#! /bin/sh -e
## DP: Adds support for MANAGESIEVE
## DP: Author: Stephan Bosch <stephan@rename-it.nl>
## DP: Version: 0.10.3

# Copyright (c) 2006-2008 by Stephan Bosch <stephan@rename-it.nl>
# This patch is licenced under LGPLv2.1 (see COPYING.LGPL in the dovecot sources)

# Patch obtained from http://www.rename-it.nl/dovecot/1.1/dovecot-1.1.1-managesieve-0.10.2.diff.gz

. $(dirname $0)/DPATCH

exit 0
@DPATCH@
diff -urN dovecot-1.1.1/dovecot-example.conf dovecot-1.1.1-managesieve-patch/dovecot-example.conf
--- dovecot-1.1.1/dovecot-example.conf	2008-06-12 00:45:10.000000000 -0600
+++ dovecot-1.1.1-managesieve-patch/dovecot-example.conf	2008-07-01 21:59:31.000000000 -0600
@@ -18,7 +18,7 @@
 # Base directory where to store runtime data.
 #base_dir = /var/run/dovecot/
 
-# Protocols we want to be serving: imap imaps pop3 pop3s
+# Protocols we want to be serving: imap imaps pop3 pop3s managesieve
 # If you only want to use dovecot-auth, you can set this to "none".
 #protocols = imap imaps
 
@@ -28,8 +28,8 @@
 # operating system. Use "*, [::]" for listening both IPv4 and IPv6.
 #
 # If you want to specify ports for each service, you will need to configure
-# these settings inside the protocol imap/pop3 { ... } section, so you can
-# specify different ports for IMAP/POP3. For example:
+# these settings inside the protocol imap/pop3/managesieve { ... } section, 
+# so you can specify different ports for IMAP/POP3/MANAGESIEVE. For example:
 #   protocol imap {
 #     listen = *:10143
 #     ssl_listen = *:10943
@@ -39,6 +39,10 @@
 #     listen = *:10100
 #     ..
 #   }
+#   protocol managesieve {
+#     listen = *:12000
+#     ..
+#   }
 #listen = *
 
 # Disable LOGIN command and all other plaintext authentications unless
@@ -650,6 +654,48 @@
 }
 
 ##
+## MANAGESIEVE specific settings
+##
+
+protocol managesieve {
+  # Login executable location.
+  #login_executable = /usr/libexec/dovecot/managesieve-login
+
+  # MANAGESIEVE executable location. See IMAP's mail_executable above for 
+  # examples how this could be changed.
+  #mail_executable = /usr/libexec/dovecot/managesieve
+
+  # Maximum MANAGESIEVE command line length in bytes. This setting is 
+  # directly borrowed from IMAP. But, since long command lines are very
+  # unlikely with MANAGESIEVE, changing this will not be very useful.  
+  #managesieve_max_line_length = 65536
+
+  # Specifies the location of the symlink pointing to the active script in
+  # the sieve storage directory. This must match the SIEVE setting used by
+  # deliver (refer to http://wiki.dovecot.org/LDA/Sieve#location for more
+  # info). Variable substitution with % is recognized.
+  sieve=~/.dovecot.sieve
+
+  # This specifies the path to the directory where the uploaded scripts must
+  # be stored. In terms of '%' variable substitution it is identical to
+  # dovecot's mail_location setting used by the mail protocol daemons.
+  sieve_storage=~/sieve
+
+  # If, for some inobvious reason, the sieve_storage remains unset, the 
+  # managesieve daemon uses the specification of the mail_location to find out 
+  # where to store the sieve files (see explaination in README.managesieve). 
+  # The example below, when uncommented, overrides any global mail_location 
+  # specification and stores all the scripts in '~/mail/sieve' if sieve_storage 
+  # is unset. However, you should always use the sieve_storage setting.
+  # mail_location = mbox:~/mail
+
+  # To fool managesieve clients that are focused on timesieved you can
+  # specify the IMPLEMENTATION capability that the dovecot reports to clients 
+  # (default: dovecot).
+  #managesieve_implementation_string = Cyrus timsieved v2.2.13
+}
+
+##
 ## LDA specific settings
 ##
 
diff -urN dovecot-1.1.1/README.managesieve dovecot-1.1.1-managesieve-patch/README.managesieve
--- dovecot-1.1.1/README.managesieve	1969-12-31 17:00:00.000000000 -0700
+++ dovecot-1.1.1-managesieve-patch/README.managesieve	2008-07-01 21:59:31.000000000 -0600
@@ -0,0 +1,29 @@
+
+MANAGESIEVE service support patch for Dovecot 1.1
+
+NOTICE: As per dovecot-1.1, this patch only contains the changes to the 
+Dovecot tree that are necessary to use the now external 
+dovecot-1.1-managesieve package. So, applying this patch is no longer 
+the only thing you need to do to give Dovecot managesieve support. 
+Eventually the need for this patch will disappear completely. 
+
+Compile
+-------
+
+You can just use the usual build process
+(http://wiki.dovecot.org/CompilingSource):
+
+./configure
+make
+sudo make install 
+
+Don't forget to execute ./autogen.sh if you downloaded dovecot from the
+Mercurial repository.
+
+Installation and Configuration
+------------------------------
+
+Read the README of the dovecot-managesieve package for the next steps. 
+Note that this patch adds configuration examples to the dovecot-example.conf 
+configuration file example.
+
diff -urN dovecot-1.1.1/src/master/child-process.c dovecot-1.1.1-managesieve-patch/src/master/child-process.c
--- dovecot-1.1.1/src/master/child-process.c	2008-06-12 00:45:10.000000000 -0600
+++ dovecot-1.1.1-managesieve-patch/src/master/child-process.c	2008-07-01 21:59:31.000000000 -0600
@@ -19,7 +19,8 @@
 	"imap",
 	"pop3",
 	"ssl-build-param",
-	"dict"
+	"dict",
+	"managesieve"
 };
 
 struct hash_table *processes;
diff -urN dovecot-1.1.1/src/master/child-process.h dovecot-1.1.1-managesieve-patch/src/master/child-process.h
--- dovecot-1.1.1/src/master/child-process.h	2008-06-12 00:45:10.000000000 -0600
+++ dovecot-1.1.1-managesieve-patch/src/master/child-process.h	2008-07-01 21:59:31.000000000 -0600
@@ -10,6 +10,7 @@
 	PROCESS_TYPE_POP3,
 	PROCESS_TYPE_SSL_PARAM,
 	PROCESS_TYPE_DICT,
+	PROCESS_TYPE_MANAGESIEVE,
 
 	PROCESS_TYPE_MAX
 };
diff -urN dovecot-1.1.1/src/master/listener.c dovecot-1.1.1-managesieve-patch/src/master/listener.c
--- dovecot-1.1.1/src/master/listener.c	2008-06-12 00:45:10.000000000 -0600
+++ dovecot-1.1.1-managesieve-patch/src/master/listener.c	2008-07-01 21:59:31.000000000 -0600
@@ -126,6 +126,10 @@
 			check_conflicts_set(server->pop3, ip, port,
 					    "pop3", proto);
 		}
+		 if (server->managesieve != NULL) {
+            check_conflicts_set(server->managesieve, ip, port,
+                        "managesieve", proto);
+        }
 	}
 }
 
@@ -226,13 +230,29 @@
 			if (set->protocol == MAIL_PROTOCOL_POP3 &&
 			    !set->ssl_disable)
 				ssl_listen = TRUE;
+		} else if (strcasecmp(*proto, "managesieve") == 0) {
+			if (set->protocol == MAIL_PROTOCOL_MANAGESIEVE)
+				nonssl_listen = TRUE;
 		}
 	}
 
 	if (!nonssl_listen)
 		listener_close_fds(&set->listens);
 	else {
-		default_port = set->protocol == MAIL_PROTOCOL_IMAP ? 143 : 110;
+		switch (set->protocol) {
+		case MAIL_PROTOCOL_IMAP:
+			default_port = 143;
+			break;
+		case MAIL_PROTOCOL_POP3:
+			default_port = 110;
+			break;
+		case MAIL_PROTOCOL_MANAGESIEVE:
+			default_port = 2000;
+			break;
+		default:
+			i_unreached();
+		}
+
 		listener_init("listen", set->listen, default_port,
 			      &set->listens);
 	}
@@ -332,9 +352,11 @@
 		if (old_set != NULL) {
 			listen_copy_old(old_set->imap, server->imap);
 			listen_copy_old(old_set->pop3, server->pop3);
+			listen_copy_old(old_set->managesieve, server->managesieve);
 		}
 		listen_parse_and_close_unneeded(server->imap);
 		listen_parse_and_close_unneeded(server->pop3);
+		listen_parse_and_close_unneeded(server->managesieve);
 
 		if (old_set != NULL)
 			old_set = old_set->next;
@@ -345,6 +367,8 @@
 			listener_listen_missing(server->imap, "imap", retry);
 		if (server->pop3 != NULL)
 			listener_listen_missing(server->pop3, "pop3", retry);
+		if (server->managesieve != NULL)
+			listener_listen_missing(server->managesieve, "managesieve", retry);
 	}
 }
 
@@ -361,5 +385,8 @@
 			listener_close_fds(&server->pop3->listens);
 			listener_close_fds(&server->pop3->ssl_listens);
 		}
+		if (server->managesieve != NULL) {
+			listener_close_fds(&server->managesieve->listens);
+		}
 	}
 }
diff -urN dovecot-1.1.1/src/master/login-process.c dovecot-1.1.1-managesieve-patch/src/master/login-process.c
--- dovecot-1.1.1/src/master/login-process.c	2008-06-12 15:38:01.000000000 -0600
+++ dovecot-1.1.1-managesieve-patch/src/master/login-process.c	2008-07-01 21:59:31.000000000 -0600
@@ -71,8 +71,20 @@
 	group = i_new(struct login_group, 1);
 	group->refcount = 1;
 	group->set = set;
-	group->mail_process_type = set->protocol == MAIL_PROTOCOL_IMAP ?
-		PROCESS_TYPE_IMAP : PROCESS_TYPE_POP3;
+
+	switch ( set->protocol ) {
+	case MAIL_PROTOCOL_IMAP:
+		group->mail_process_type = PROCESS_TYPE_IMAP;
+		break;
+	case MAIL_PROTOCOL_POP3:
+		group->mail_process_type = PROCESS_TYPE_POP3;
+		break;
+	case MAIL_PROTOCOL_MANAGESIEVE:
+		group->mail_process_type = PROCESS_TYPE_MANAGESIEVE;
+		break;
+	default:
+		i_unreached();
+	}
 
 	group->next = login_groups;
 	login_groups = group;
@@ -224,6 +236,8 @@
 			login_group_create(server->imap);
 		if (server->pop3 != NULL)
 			login_group_create(server->pop3);
+		if (server->managesieve != NULL)
+			login_group_create(server->managesieve);
 	}
 }
 
@@ -285,6 +299,8 @@
 			protocol = MAIL_PROTOCOL_IMAP;
 		else if (strcmp(proto, "pop3") == 0)
 			protocol = MAIL_PROTOCOL_POP3;
+		else if (strcmp(proto, "managesieve") == 0)
+			protocol = MAIL_PROTOCOL_MANAGESIEVE;
 		else {
 			i_error("login: Unknown protocol '%s'", proto);
 			return FALSE;
@@ -579,7 +595,10 @@
 				    *set->imap_capability != '\0' ?
 				    set->imap_capability :
 				    set->imap_generated_capability, NULL));
-	}
+	} else if (group->mail_process_type == PROCESS_TYPE_MANAGESIEVE) {
+		env_put(t_strconcat("MANAGESIEVE_IMPLEMENTATION_STRING=",
+			set->managesieve_implementation_string, NULL));
+    }
 }
 
 static pid_t create_login_process(struct login_group *group)
diff -urN dovecot-1.1.1/src/master/mail-process.c dovecot-1.1.1-managesieve-patch/src/master/mail-process.c
--- dovecot-1.1.1/src/master/mail-process.c	2008-06-12 00:45:10.000000000 -0600
+++ dovecot-1.1.1-managesieve-patch/src/master/mail-process.c	2008-07-01 21:59:31.000000000 -0600
@@ -329,6 +329,27 @@
 	/* We care about POP3 UIDL format in all process types */
 	env_put(t_strconcat("POP3_UIDL_FORMAT=", set->pop3_uidl_format, NULL));
 
+	if (set->protocol == MAIL_PROTOCOL_MANAGESIEVE) {
+		env_put(t_strdup_printf("MANAGESIEVE_MAX_LINE_LENGTH=%u",
+				    set->managesieve_max_line_length));
+		env_put(t_strconcat("MANAGESIEVE_IMPLEMENTATION_STRING=",
+        		    set->managesieve_implementation_string, NULL));
+	}
+
+	/* Set sieve environment 
+	 *   FIXME: Currently just uses the expand_mail_env function to 
+	 *   substitute variables and home dir. Technically, that function
+	 *   is not meant for the sieve implementation.  
+	 */
+	if ( set->sieve_storage != NULL ) {
+		env_put(t_strconcat("SIEVE_STORAGE=",
+        	expand_mail_env(set->sieve_storage, var_expand_table), NULL));
+	}
+	if (set->sieve != NULL) {
+		env_put(t_strconcat("SIEVE=",
+			expand_mail_env(set->sieve, var_expand_table), NULL));
+	}
+
 	if (set->mail_save_crlf)
 		env_put("MAIL_SAVE_CRLF=1");
 	if (set->mmap_disable)
@@ -441,6 +462,8 @@
 			set = server->imap;
 		else if (strcmp(protocol, "pop3") == 0)
 			set = server->pop3;
+		else if (strcmp(protocol, "managesieve") == 0)
+			set = server->managesieve;
 		else
 			i_fatal("Unknown protocol: '%s'", protocol);
 		executable = set->mail_executable;
@@ -540,7 +563,8 @@
 	bool home_given, nfs_check;
 
 	i_assert(process_type == PROCESS_TYPE_IMAP ||
-		 process_type == PROCESS_TYPE_POP3);
+		process_type == PROCESS_TYPE_POP3 || 
+		process_type == PROCESS_TYPE_MANAGESIEVE );
 
 	if (mail_process_count == set->max_mail_processes) {
 		i_error("Maximum number of mail processes exceeded");
@@ -895,6 +919,8 @@
 					   mail_process_destroyed);
 	child_process_set_destroy_callback(PROCESS_TYPE_POP3,
 					   mail_process_destroyed);
+	child_process_set_destroy_callback(PROCESS_TYPE_MANAGESIEVE,
+					   mail_process_destroyed);
 }
 
 void mail_processes_deinit(void)
diff -urN dovecot-1.1.1/src/master/main.c dovecot-1.1.1-managesieve-patch/src/master/main.c
--- dovecot-1.1.1/src/master/main.c	2008-06-12 00:45:10.000000000 -0600
+++ dovecot-1.1.1-managesieve-patch/src/master/main.c	2008-07-01 21:59:31.000000000 -0600
@@ -181,6 +181,8 @@
 			return TRUE;
 		if (server->pop3 != NULL && have_stderr_set(server->pop3))
 			return TRUE;
+		if (server->managesieve != NULL && have_stderr_set(server->managesieve))
+			return TRUE;
 
 		server = server->next;
 	}
diff -urN dovecot-1.1.1/src/master/master-settings.c dovecot-1.1.1-managesieve-patch/src/master/master-settings.c
--- dovecot-1.1.1/src/master/master-settings.c	2008-06-21 07:09:16.000000000 -0600
+++ dovecot-1.1.1-managesieve-patch/src/master/master-settings.c	2008-07-01 21:59:31.000000000 -0600
@@ -293,6 +293,14 @@
 	/* dict */
 	MEMBER(dict_db_config) NULL,
 
+	/* managesieve */
+	MEMBER(managesieve_max_line_length) 65536,
+	MEMBER(managesieve_implementation_string) PACKAGE,
+
+	/* sieve */
+	MEMBER(sieve_storage) "",
+	MEMBER(sieve) NULL,
+
 	/* .. */
 };
 
@@ -473,6 +481,8 @@
 			auth->parent->pop3->ssl_verify_client_cert = TRUE;
 		if (auth->parent->imap != NULL)
 			auth->parent->imap->ssl_verify_client_cert = TRUE;
+		if (auth->parent->managesieve != NULL)
+			auth->parent->managesieve->ssl_verify_client_cert = TRUE;
 	}
 
 	for (s = auth->sockets; s != NULL; s = s->next) {
@@ -523,9 +533,12 @@
 	if (set->protocol == MAIL_PROTOCOL_IMAP) {
 		if (strstr(set->protocols, "imap") == NULL)
 			return FALSE;
-	} else {
+	} else if (set->protocol == MAIL_PROTOCOL_POP3) {
 		if (strstr(set->protocols, "pop3") == NULL)
 			return FALSE;
+	} else {
+		if (strstr(set->protocols, "managesieve") == NULL)
+			return FALSE;
 	}
 
 	return TRUE;
@@ -1138,7 +1151,7 @@
 
 	if (strcmp(key, "login") == 0) {
 		i_warning("Ignoring deprecated 'login' section handling. "
-			  "Use protocol imap/pop3 { .. } instead. "
+			  "Use protocol imap/pop3/managesieve { .. } instead. "
 			  "Some settings may have been read incorrectly.");
 		return NULL;
 	}
@@ -1164,6 +1177,15 @@
 							key, value);
 		}
 
+		if (error == NULL &&
+		    (ctx->protocol == MAIL_PROTOCOL_ANY ||
+		     ctx->protocol == MAIL_PROTOCOL_MANAGESIEVE)) {
+			error = parse_setting_from_defs(settings_pool,
+							setting_defs,
+							ctx->server->managesieve,
+							key, value);
+		}
+
 		if (error == NULL)
 			return NULL;
 
@@ -1222,6 +1244,13 @@
 			array_append(&ctx->server->pop3->plugin_envs,
 				     &value, 1);
 		}
+		if (ctx->protocol == MAIL_PROTOCOL_ANY ||
+		    ctx->protocol == MAIL_PROTOCOL_MANAGESIEVE) {
+			array_append(&ctx->server->managesieve->plugin_envs, &key, 1);
+			array_append(&ctx->server->managesieve->plugin_envs,
+				     &value, 1);
+		}
+
 		return NULL;
 	}
 
@@ -1231,7 +1260,8 @@
 static struct server_settings *
 create_new_server(const char *name,
 		  struct settings *imap_defaults,
-		  struct settings *pop3_defaults)
+		  struct settings *pop3_defaults,
+		  struct settings *managesieve_defaults)
 {
 	struct server_settings *server;
 
@@ -1239,14 +1269,17 @@
 	server->name = p_strdup(settings_pool, name);
 	server->imap = p_new(settings_pool, struct settings, 1);
 	server->pop3 = p_new(settings_pool, struct settings, 1);
+	server->managesieve = p_new(settings_pool, struct settings, 1);
 	server->auth_defaults = default_auth_settings;
 
 	*server->imap = *imap_defaults;
 	*server->pop3 = *pop3_defaults;
+	*server->managesieve = *managesieve_defaults;
 
 	p_array_init(&server->dicts, settings_pool, 4);
 	p_array_init(&server->imap->plugin_envs, settings_pool, 8);
 	p_array_init(&server->pop3->plugin_envs, settings_pool, 8);
+	p_array_init(&server->managesieve->plugin_envs, settings_pool, 8);
 
 	server->imap->server = server;
 	server->imap->protocol = MAIL_PROTOCOL_IMAP;
@@ -1260,6 +1293,12 @@
 	server->pop3->mail_executable = PKG_LIBEXECDIR"/pop3";
 	server->pop3->mail_plugin_dir = MODULEDIR"/pop3";
 
+	server->managesieve->server = server;
+	server->managesieve->protocol = MAIL_PROTOCOL_MANAGESIEVE;
+	server->managesieve->login_executable = PKG_LIBEXECDIR"/managesieve-login";
+	server->managesieve->mail_executable = PKG_LIBEXECDIR"/managesieve";
+	server->managesieve->mail_plugin_dir = MODULEDIR"/managesieve";
+  
 	return server;
 }
 
@@ -1302,8 +1341,8 @@
 
 		ctx->type = SETTINGS_TYPE_SERVER;
 		ctx->server = create_new_server(name, ctx->server->imap,
-						ctx->server->pop3);
-                server = ctx->root;
+						ctx->server->pop3, ctx->server->managesieve);
+		server = ctx->root;
 		while (server->next != NULL)
 			server = server->next;
 		server->next = ctx->server;
@@ -1324,6 +1363,8 @@
 			ctx->protocol = MAIL_PROTOCOL_POP3;
 		else if (strcmp(name, "lda") == 0)
 			ctx->protocol = MAIL_PROTOCOL_LDA;
+		else if (strcmp(name, "managesieve") == 0)
+			ctx->protocol = MAIL_PROTOCOL_MANAGESIEVE;
 		else {
 			*errormsg = "Unknown protocol name";
 			return FALSE;
@@ -1435,6 +1476,8 @@
 			fd_count += server->imap->login_max_processes_count;
 		if (server->pop3 != NULL)
 			fd_count += server->pop3->login_max_processes_count;
+		if (server->managesieve != NULL)
+			fd_count += server->managesieve->login_max_processes_count;
 		fd_count += server->defaults->max_mail_processes;
 	}
 
@@ -1464,7 +1507,7 @@
 	ctx.protocol = MAIL_PROTOCOL_ANY;
 	ctx.server = ctx.root =
 		create_new_server("default",
-				  &default_settings, &default_settings);
+				  &default_settings, &default_settings, &default_settings);
 	ctx.auth = &ctx.server->auth_defaults;
 
 	if (!settings_read(path, NULL, parse_setting, parse_section, &ctx))
@@ -1481,7 +1524,9 @@
 
 	if (!nochecks && !nofixes) {
 		ctx.root->defaults = settings_is_active(ctx.root->imap) ?
-			ctx.root->imap : ctx.root->pop3;
+			ctx.root->imap : 
+			(settings_is_active(ctx.root->pop3) ? 
+				ctx.root->pop3 : ctx.root->managesieve);
 
 		path = t_strconcat(ctx.root->defaults->base_dir,
 				   "/master.pid", NULL);
@@ -1491,7 +1536,8 @@
 	prev = NULL;
 	for (server = ctx.root; server != NULL; server = server->next) {
 		if ((*server->imap->protocols == '\0' ||
-		     *server->pop3->protocols == '\0') && !nochecks) {
+		     *server->pop3->protocols == '\0' ||
+		     *server->managesieve->protocols == '\0') && !nochecks) {
 			i_error("No protocols given in configuration file");
 			return FALSE;
 		}
@@ -1521,6 +1567,15 @@
 				server->defaults = server->pop3;
 		}
 
+		if (!settings_is_active(server->managesieve) && !nochecks)
+			server->managesieve = NULL;
+		else {
+			if (!settings_fix(server->managesieve, nochecks, nofixes))
+				return FALSE;
+			if (server->defaults == NULL)
+				server->defaults = server->managesieve;
+		}
+
 		if (server->defaults == NULL) {
 			if (prev == NULL)
 				ctx.root = server->next;
@@ -1738,8 +1793,8 @@
 
 void master_settings_dump(struct server_settings *set, bool nondefaults)
 {
-	const void *sets[4];
-	const char *set_names[4];
+	const void *sets[5];
+	const char *set_names[5];
 	unsigned int count;
 
 	sets[0] = &default_settings;
@@ -1758,6 +1813,11 @@
 		sets[count] = set->pop3;
 		set_names[count] = "pop3";
 		count++;
+	}	
+	if (set->managesieve != NULL) {
+		sets[count] = set->managesieve;
+		set_names[count] = "managesieve";
+		count++;
 	}
 	settings_dump(setting_defs, sets, set_names, count, nondefaults, 0);
 	namespace_settings_dump(set->namespaces, nondefaults);
diff -urN dovecot-1.1.1/src/master/master-settings-defs.c dovecot-1.1.1-managesieve-patch/src/master/master-settings-defs.c
--- dovecot-1.1.1/src/master/master-settings-defs.c	2008-06-12 00:45:10.000000000 -0600
+++ dovecot-1.1.1-managesieve-patch/src/master/master-settings-defs.c	2008-07-01 21:59:31.000000000 -0600
@@ -127,5 +127,13 @@
 	/* dict */
 	DEF_STR(dict_db_config),
 
+  	/* managesieve */
+  	DEF_INT(managesieve_max_line_length),
+  	DEF_STR(managesieve_implementation_string),
+
+	/* sieve */
+  	DEF_STR(sieve_storage),
+  	DEF_STR(sieve),
+
 	{ 0, NULL, 0 }
 };
diff -urN dovecot-1.1.1/src/master/master-settings.h dovecot-1.1.1-managesieve-patch/src/master/master-settings.h
--- dovecot-1.1.1/src/master/master-settings.h	2008-06-12 00:45:10.000000000 -0600
+++ dovecot-1.1.1-managesieve-patch/src/master/master-settings.h	2008-07-01 21:59:31.000000000 -0600
@@ -4,10 +4,11 @@
 #include "network.h"
 
 enum mail_protocol {
-        MAIL_PROTOCOL_ANY,
-        MAIL_PROTOCOL_IMAP,
+	MAIL_PROTOCOL_ANY,
+	MAIL_PROTOCOL_IMAP,
 	MAIL_PROTOCOL_POP3,
-	MAIL_PROTOCOL_LDA
+	MAIL_PROTOCOL_LDA,
+	MAIL_PROTOCOL_MANAGESIEVE
 };
 
 struct listener {
@@ -139,6 +140,14 @@
 	/* dict */
 	const char *dict_db_config;
 
+	/* managesieve */
+	unsigned int managesieve_max_line_length;
+	const char *managesieve_implementation_string;
+
+	/* sieve */
+	const char *sieve_storage;
+	const char *sieve;
+
 	/* .. */
 	ARRAY_TYPE(listener) listens;
 	ARRAY_TYPE(listener) ssl_listens;
@@ -252,9 +261,10 @@
 	struct settings *defaults;
 	struct settings *imap;
 	struct settings *pop3;
+	struct settings *managesieve;
 	struct auth_settings *auths;
 	struct auth_settings auth_defaults;
-        struct namespace_settings *namespaces;
+	struct namespace_settings *namespaces;
 
 	ARRAY_DEFINE(dicts, const char *);
 
